_overviews/scala3-book/tasty-for-end-users.md

```diff
@@ -0,0 +1,150 @@
+
+```scala
+val x = xs(0)
+```
+
+the resulting class file will have a casting operation for this line of code, which you can think of being like this:
+
+```
+int x = (Int) xs.get(0)               // Java-ish
+val x = xs.get(0).asInstanceOf[Int]   // more Scala-like
+```
+
+Again, this is done for compatibility, so your Scala code can run on the JVM.
+
+And this discussion only covers the topic of type erasure. There are similar issues for every other Scala construct that the JVM isn’t aware of, including constructs like unions, intersections, traits with parameters, and many more Scala 3 features.
+
+
+### Key points
+
+So that’s the first takeaway from this section: The types you specify in your Scala code aren’t represented completely accurately in _.class_ files.
+
+A second key point is to understand that there are differences between the information that’s available at _compile time_ and _run time_:
+
+- At **compile time**, as `scalac` reads and analyzes your code, it knows that `xs` is a `List[Int]`
+- When the compiler writes your code to a class file, it writes that `xs` is a `List[Object]`, and it adds casting information everywhere else `xs` is accessed
+- Then at **run time** --- with your code running inside the JVM --- the JVM doesn’t know that your list is a `List[Int]`
+
+With Scala 3 and Tasty, here’s another important note about compile time:
+
+- When you write Scala 3 code that uses other Scala 3 libraries, `scalac` doesn’t have to read their _.class_ files any more; it can read their _.tasty_ files, which, as mentioned, are an _exact_ representation of your code. This helps in many ways.
+
+{% comment %}
+NOTE: In regards to the previous paragraphs, I’ve read that there’s also metadata in the class files, but I don’t know those details.
+{% endcomment %}
+
+
+
+## Tasty benefits
+
+As you can imagine, having a complete representation of your code in a platform-neutral form has many benefits:
+
+{% comment %}
+TODO: These lines mostly come from https://www.scala-lang.org/blog/2018/04/30/in-a-nutshell.html, and I don’t know exactly what the first bullet point means:
+{% endcomment %}
+
+- The compiler uses it to support separate compilation.
+- The Scala _Language Server Protocol_-based language server uses it to support hyperlinking, command completion, documentation, and also for global operations such as find-references and renaming.
+- Tasty makes an excellent foundation for a new generation of reflection-based macros.
+- A build tool can use it to cross-build on different platforms and migrate code from one binary version to another.
+- Optimizers and analyzers can use it for deep code analysis and advanced code generation.
+
+There’s one more big benefit:
+
+- Because the Scala 2.13.4 compiler can read Tasty files, and the Scala 3 compiler can read Scala 2.13 class files, you can intermix Scala 3 and Scala 2.13 code in the same project.
```

Comment:
This is not really an advantage of Tasty. Is Scala 3 had used pickles like Scala 2, even with a different format, we would still have been able to do this. The fact that we use tasty for this is simply because that's where the information is, not because tasty is any better at storing that information than what we had before.

Reply:
Okay, thanks, I’ll delete this.